refactor: change pydantic dataclass decorator

PrettyWood opened the pull request:
<!-- Thank you for your contribution! -->
<!-- Unless your change is trivial, please create an issue to discuss the change before creating a PR -->
<!-- See https://pydantic-docs.helpmanual.io/contributing/ for help on Contributing -->

## Change Summary
Now this is doable

```py
from dataclasses import dataclass as stdlib_dataclass

from pydantic.dataclasses import dataclass as validated_dataclass


@stdlib_dataclass(frozen=True)
class Item:
    name: str
    b: int


ValidatedItem = validated_dataclass(Item)


item = Item(name='qwe', b=1)
item_b = Item(name='qwe', b=b'1')

item_2 = ValidatedItem(name='qwe', b=1)
item_b2 = ValidatedItem(name='qwe', b=b'1')

assert item != item_b
assert item == item_2 == item_b2

assert hash(item) != hash(item_b)
assert hash(item) == hash(item_2) == hash(item_b2)

assert item_b == ValidatedItem(name='qwe', b=b'1', __pydantic_run_validation__=False)
```

<!-- Please give a short summary of the changes. -->

## Related issue number
A bunch on stdlib dataclass to validated dataclass
closes #2162
closes #2383 
closes #2398
closes #2424
closes #2541
closes #2555
closes #2594
closes #2511
closes #3011
closes #3046
closes #3109
closes #3162
closes #3709

<!-- Are there any issues opened that will be resolved by merging this change? -->

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

samuelcolvin commented:
Looks interesting, why `dataclasses2`? Also, how much has changed from the `pydantic.dataclasses` module?

PrettyWood commented:
@samuelcolvin The `dataclasses2` was just in the meantime to showcase the new proposal of implementation. There is still some cleanup to do but some new bases are here and I added all the tests from the 5 issues 🤯 
I'm afraid I'll still need to find another workaround as we need our proxy to forward also inheritance, which is done thanks to `__mro_entries__` available since....3.7!
I'll keep digging but if anyone has an idea I'm all ears 😅

PrettyWood commented:
@samuelcolvin The `dataclasses2` was just in the meantime to showcase the new proposal of implementation. There is still some cleanup to do but some new bases are here and I added all the tests from the 6 issues 🤯

I'm afraid we have a small problem because we need the proxy to forward inheritance, which is done thanks to `__mro_entries__` available since....3.7! 😓 

Remember that the proxy solution seems like a good idea because we need to update in place the stdlib dataclass to support validation and still have all comparisons between std and validated dataclasses.
BUT we still want to have the following behaviour
```py
@dataclasses.dataclass
class M:
    x: str

ValidatedM = pydantic.dataclasses.dataclass(M)

assert M(x=123).x == 123
assert ValidatedM(x=123).x == '123'
```

-----------------------------------------

### 4 options:

1) support 3.7+ only for pydantic dataclass

2) Change in place and probably avoid side effects on stdlib dataclass
<details>
  <summary>Click to expand!</summary>
  
```python
@dataclasses.dataclass
class DC:
    x: int

assert DC(x='3').x == '3'

ValidatedDC = pydantic.dataclasses.dataclass(DC)
# the decorator will add every magic methods in place but not alter default behaviour of stdlib dataclass
assert ValidatedDC(x='3').x == '3'
assert ValidatedDC(x='3', __pydantic_run_validation__=True).x == 3
```
and validate by default if pydantic decorator is used directly

```python
@dataclasses.dataclass
class ValidatedDC:
    x: int

assert ValidatedDC(x='3').x == 3
assert ValidatedDC(x='3', __pydantic_run_validation__=False).x == '3'
```

I'm just not perfectly happy with the difference between plain decorator and decorator of stdlib one
</details>

3) A mix of 1) and 2) (1 for 3.7+, degraded experience for 3.6 but IMO it's acceptable)

4) A genius idea from someone else? I'm all ears!! 😅

Personally since I don't have other solutions in mind for now I would be in favor of 3 since I'm afraid there are still people using 3.6 + dataclass backport and we don't want to break something that has just been added in 1.7 😐

PrettyWood commented:
@samuelcolvin The `dataclasses2` was just in the meantime to showcase the new proposal of implementation.

So I'm pretty happy with the current state of this PR
I added all the tests from the 6 issues but plenty more 🤯

The chosen behaviour is
```py
import dataclasses
import pydantic


@pydantic.dataclasses.dataclass
class M1:
    x: int


@pydantic.dataclasses.dataclass
@dataclasses.dataclass
class M2:
    x: int


@dataclasses.dataclass
class M3:
    x: int

VM3 = pydantic.dataclasses.dataclass(M3)

assert M1(x='3').x == 3
assert M2(x='3').x == 3
assert M3(x=3).x == 3
assert VM3(x='3').x == 3
```

The only issue is for the 3rd case, which should be quite rare!
To do that I added a proxy (taken from the great [wrapt](https://github.com/GrahamDumpleton/wrapt) library). But to work properly we need the proxy to forward inheritance, which is done thanks to `__mro_entries__` available since....3.7! 😓 

For python 3.6 we hence tolerate a degraded behaviour with a side effect on the stdlib dataclass: it now triggers validation by default. But it can be changed directly in the validator on thanks to a magic kwarg in `__init__`. The user is naturally notified with a `UserWarning`!
(It will give even more reasons for people to move to a more recent python version 😆)

@samuelcolvin I wait for your remarks before updating the documentation. It's quite heavy sorry for that!

antonl commented:
@PrettyWood another testcase to add to your branch #2594

samuelcolvin commented on pydantic/wrapper.py:
```diff
@@ -0,0 +1,409 @@
+"""
```
humm, is this really necessary, can't we implement a few methods like `__repr__`, `__eq__` etc. and get the all the functionality required?

This seems like it might confuse people (e.g. in trackbacks) and damage performance.

samuelcolvin replied:
if we do need this, please could you add an `__all__`?

samuelcolvin commented:
please update.

PrettyWood commented on pydantic/dataclasses.py:
```diff
@@ -32,136 +77,217 @@ def __get_validators__(cls: Type['Dataclass']) -> 'CallableGenerator':
+
+        @wraps(init)
+        def new_init(
+            self: 'Dataclass', *args: Any, __pydantic_run_validation__: bool = validate_on_init, **kwargs: Any
```
You're totally right! Nice catch thanks! This is actually a big thing. I'll work on other PRs before coming back on this one.

PrettyWood commented on pydantic/dataclasses.py:
```diff
@@ -32,136 +77,217 @@ def __get_validators__(cls: Type['Dataclass']) -> 'CallableGenerator':
+    dc_cls: Type['Dataclass'],
+    config: Optional[Type['BaseConfig']],
+    validate_on_init: bool,
+    dc_cls_doc: Optional[str] = None,
```
It's always provided but is `None` by default for a plain class

PrettyWood commented:
@samuelcolvin Ok I simplified a lot the logic of the wrapper (just kept `__getattr__`, which is enough for all bugs). I also put the logic to trigger validation as class attribute (instead of `__init__` parameter), which seems safer and in fact way easier (removes all the `partial` logic for `__post_init__`).
Note that the extra overhead of the wrapper may slow things down a little but it's only if people use pydantic dataclass on top of a stdlib one. And maybe we can add a parameter to avoid this logic and accept the side effect on stdlib dataclass ❓ 
Thanks for the feedback! Please review again 😬 🙏

ahirner reviewed (commented):
I like the new and concise approach.

That said, I wouldn't use the `pydantic_dataclass(User)` feature. They would often be types from other libraries and I'd avoid mutating them. 

Instead, I'd derive from such classes and expect a new python subtype which is mutated into whatever it needs to be. 

As of now, the wrapper proxy is returned. I don't see the benefit of using it in general. Maybe I'm missing some constraint that makes it required?

```python
from dataclasses import dataclass
from pydantic.dataclasses import dataclass as p_dataclass, is_builtin_dataclass

@dataclass
class X:
    a: int

@p_dataclass
class Y(X):
    pass

# passes
assert type(X("ohno")) == X, f"expected X to by X not {X}"

# doesn't
assert type(Y(10)) == Y, f"expected Y to by Y not {Y}"
```

PrettyWood commented on pydantic/dataclasses.py:
```diff
@@ -32,136 +75,206 @@ def __get_validators__(cls: Type['Dataclass']) -> 'CallableGenerator':
+    """
+
+    def wrap(cls: Type[Any]) -> 'DataclassClassOrWrapper':
         import dataclasses
```
For import time of the whole `pydantic` module https://github.com/samuelcolvin/pydantic/pull/1132

PrettyWood commented on pydantic/dataclasses.py:
```diff
@@ -32,136 +75,206 @@ def __get_validators__(cls: Type['Dataclass']) -> 'CallableGenerator':
-    This is why we also add a class attribute `__processed__` to only consider 'direct' built-in dataclasses
+    We need to replace the right method. If no `__post_init__` has been set in the stdlib dataclass
+    it won't even exist (code is generated on the fly by `dataclasses`)
+    By default, we run validation after `__init__` or `__post_init__` if defined
```
yep (see `__post_init_post_parse__`)

PrettyWood replied:
Btw I added https://github.com/samuelcolvin/pydantic/pull/2557/commits/d1014201aa32f83f5bb8aecce8912145f2232f8c for some cases where you have sdtlib dataclass

PrettyWood commented on pydantic/dataclasses.py:
```diff
@@ -176,92 +289,65 @@ def _process_class(
+    return (
+        not hasattr(_cls, '__processed__')
+        and dataclasses.is_dataclass(_cls)
+        and set(_cls.__dataclass_fields__).issuperset(set(getattr(_cls, '__annotations__', {})))
```
I would need to go back into the code! 😅 (I will try in September)
In the meantime you're welcome to try out and if you find smarter ways that don't break the cases, I'll be more than happy to add them 😄

PrettyWood commented:
> That said, I wouldn't use the pydantic_dataclass(User) feature. They would often be types from other libraries and I'd avoid mutating them. Instead, I'd derive from such classes and expect a new python subtype which is mutated into whatever it needs to be.

Well we used to create a subclass but it's not always expected. If you read all the linked issues, you may see all the constraints. Some people want a standard dataclass used by pydantic to trigger validation but still behave as expected (comparison, hash, ...)
Using directly a subclass is not the answer (but maybe I missed something it's clearly possible!)
Hence the wrapper class (which is actually a subclass...but that also forwards inheritance, comparison, ...)

PrettyWood commented:
The CI error on fastapi side is expected
```diff
--- a/tests/test_tutorial/test_dataclasses/test_tutorial002.py
+++ b/tests/test_tutorial/test_dataclasses/test_tutorial002.py
@@ -29,7 +29,7 @@ openapi_schema = {
         "schemas": {
             "Item": {
                 "title": "Item",
-                "required": ["name", "price", "tags"],
+                "required": ["name", "price"],
                 "type": "object",
                 "properties": {
                     "name": {"title": "Name", "type": "string"},
```
`tags` should not be required since it has `default_factory=list`

PrettyWood commented:
<img width="883" alt="Screen Shot 2021-09-07 at 2 21 02 AM" src="https://user-images.githubusercontent.com/18406791/132267598-66f03738-77a5-4458-b4ae-cd59e62637b9.png">

robons commented on pydantic/dataclasses.py:
```diff
@@ -33,136 +76,210 @@ def __get_validators__(cls: Type['Dataclass']) -> 'CallableGenerator':
+    def __init__(self, dc_cls: Type['Dataclass']) -> None:
+        object.__setattr__(self, '__dataclass__', dc_cls)
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Any:
```
I guess we lose the intellisense/ide autocompletion for the underlying class init parameters in addition to static type checking with this approach, don't we?

robons replied:
Also, do we lose `isinstance(wrapper_instance,wrapper_instance.__dataclass__.__class__)` with this proxy?

PrettyWood replied:
Yep currently we lose intellisense. I tried to have a `if TYPE_CHECKING` block with a generic but didn't see anything in typing to mimic the `__call__` to the  one of `T`.
If you have any clue feel free to add it!

For the second question this is what we have
```py
from dataclasses import dataclass as sdlib_dataclass

from pydantic.dataclasses import dataclass as py_dataclass


@sdlib_dataclass
class A:
    x: int

a = A(x=1)

A_PY = py_dataclass(A)
validated_a = A_PY(x='1')

assert validated_a.x == 1
assert isinstance(validated_a, A)
assert not isinstance(validated_a, A_PY)
```

samuelcolvin replied:
mypy doesn't recognising a pydantic dataclass as dataclass with this, but from my tests it doesn't on master/v1.8 either.

It would be great if there was a way to get this to work out of the box. But since it's not a regression, it shouldn't block this PR.

samuelcolvin commented on pydantic/dataclasses.py:
```diff
@@ -33,136 +76,210 @@ def __get_validators__(cls: Type['Dataclass']) -> 'CallableGenerator':
+
+    def __getattr__(self, name: str) -> Any:
+        return getattr(self.__dataclass__, name)
+
```
```suggestion

    def __instancecheck__(self, instance):
        return isinstance(instance, self.__dataclass__)

```

Should fix the `assert not isinstance(validated_a, A_PY)` case in the above comment.

samuelcolvin reviewed (commented):
This is looking great.

But I think it's causing failures in FastAPI. Maybe @tiangolo could look into this?

samuelcolvin commented:
Please update.

Also, I'm a big confused about where we are with mypy compatibility. Does this make things better or worse, or not really make a difference.

If it makes things worse in any way, we should probably make that clear.

PrettyWood commented:
@samuelcolvin I think it makes things a bit worse but honestly pydantic dataclasses don't fit well with `mypy` since day 1. The plugin needs some work. I can still write some documentation on it.
As for `fastapi` I'll try to get @tiangolo on twitter. The test that is failing is actually a mistake that this PR fixes

PrettyWood commented:
please review 🙏

tiangolo commented:
Amazing, thanks for all your work @PrettyWood! 🚀🙇

I just merged your PR in FastAPI and made a FastAPI release (`0.70.1`) just to unblock you here (so that tests can pick the latest FastAPI release). I just manually re-ran all the jobs here and it seems they are all passing now. 🎉

samuelcolvin commented on changes/2557-PrettyWood.md:
```diff
@@ -0,0 +1,9 @@
+Finally, config customization directly via a `dict` is now possible.
+<br/><br/>
+**BREAKING CHANGES**
+- The `compiled` boolean (whether _pydantic_ is compiled with cython) has been moved from `main.py` to `version.py`
```
does this really count as breaking? The import from main will still work.

PrettyWood replied:
`from pydantic.main import compiled` will fail

samuelcolvin reviewed (commented):
Looking good, but still some stuff to fix.

We also need a warning in the docs about `config.extra`,

Taking this example:

```py
from pydantic import ConfigDict
from pydantic.dataclasses import dataclass


@dataclass(config=ConfigDict(extra='allow'))
class Foobar:
    a: int
    b: str

    # def __post_init__(self):
    #     print(self.a, self.b)


f = Foobar(a=1, b='2', c=4)
debug(f, f.c)
```
We get the following, :
* with `__post_init__` commented out: validation works, `print(f)` does not include `c` but `f.c` is right
* with `__post_init__` included: validation fails

We need a warning about `print(dc)` and `dataclasses.fields(dc)`  not including the extra properties, and we either need a prominent warning about `extra with `__post_init__` or to fix it.

samuelcolvin commented:
please update.

PrettyWood commented on pydantic/dataclasses.py:
```diff
@@ -33,136 +76,232 @@ def __get_validators__(cls: Type['Dataclass']) -> 'CallableGenerator':
-        raise DataclassTypeError(class_name=cls.__name__)
+        if is_builtin_dataclass(cls):
+            should_validate_on_init = False if validate_on_init is None else validate_on_init
+            _add_pydantic_validation_attributes(cls, the_config, should_validate_on_init, '')
```
fb77d0d

PrettyWood commented on pydantic/dataclasses.py:
```diff
@@ -33,136 +76,232 @@ def __get_validators__(cls: Type['Dataclass']) -> 'CallableGenerator':
-        _cls, init=init, repr=repr, eq=eq, order=order, unsafe_hash=unsafe_hash, frozen=frozen
-    )
-    cls.__processed__ = ClassAttribute('__processed__', True)
+                def ignore_extra_init(self: 'Dataclass', *a: Any, **kw: Any) -> None:
```
Thanks! Was I drunk or something? 1a60ab2

PrettyWood commented on pydantic/config.py:
```diff
@@ -100,6 +143,25 @@ def prepare_field(cls, field: 'ModelField') -> None:
+        return BaseConfig
+
+    else:
+        config_dict = (
```
The thing is most people write
```
class Config:
 ....
```
without actually setting `BaseConfig` as the parent.

PrettyWood commented:
@samuelcolvin 
For https://github.com/samuelcolvin/pydantic/pull/2557#discussion_r771843601 I looked at it again and explained (and also cleaned up) in 4ce2454

PrettyWood commented:
@samuelcolvin I took most remarks into consideration.
I feel like having this in 1.9 would be great and I'm willing to put more effort into this if needed.
Clearly it's not yet perfect but it's a nice step forward and it fixes already plenty of issues.
I'm just waiting for your feedback on some concerns.
Feel free to update yourself when you review if you have time.
Please review 🙏

DetachHead commented:
when using `Extra.allow` the extra properties don't show up in `__str__`
```py
from pydantic import Extra
from pydantic.dataclasses import dataclass


class Config:
    extra = Extra.allow


@dataclass(config=Config)
class Foo:
    a: int


foo = Foo(a=1, b=2)

print(foo.b) # 2
print(foo) # Foo(a=1)
```

ahirner commented:
> when using `Extra.allow` the extra properties don't show up in `__str__`

@DetachHead This caveat is documented in https://github.com/samuelcolvin/pydantic/pull/2557/commits/5907f8207a6e0c0810905b6af9b56ba5fa1db4f9

DetachHead commented:
is there an issue tracking it? if not i can make one

haf commented:
Can we please merge this and cut a release? :)

samuelcolvin commented:
Sorry, been busy with pydantic-core, I'll work on 1.10 soon and this will be included.

@PrettyWood what do you think?

PrettyWood commented:
@samuelcolvin taking some days off after a very busy Q2! If you can resolve conflicts, I reckon it's solving a lot of issues with dataclasses. And we'll probably have cleaner logic (and hence code) in v2

samuelcolvin commented:
This is now just conflicting with #3713, I'm not sure why, I thought I sorted it. Any suggestions? Otherwise I'll dig deeper.

samuelcolvin commented:
Thank you so much @PrettyWood :tada:, closing 10 issues with one PR must be a record.

Sorry everyone for the delay on this.

PoojaAg18 commented:
ValidatedItem(name='qwe', b=b'1', __pydantic_run_validation__=False). using __pydantic_run_validation__ doesn't seems to skip validation. I am using pydantic  -1.10.2 @samuelcolvin @PrettyWood
